### DISCUSSION TITLE
Have ability to order Mobjects on scene by animate method

### DISCUSSION THREAD

**Discussion by ponmike92** [ManimCommunity/manim#2778] [Suggestions and Proposals]
@kolibril13 had already mentioned some kind of that request in 3b1b's Manim here https://github.com/3b1b/manim/issues/852.

Sorry If duplicate. I didn't find it in ManimCE discussions.

## Main idea
Have ability to:
1. Order objects closer or further to each other on 2D scene.
2. Change that order DURING animation.

It can be reached, for example, comparing mobjects z_indices. @kolibril13 thanks for image

![z-order](https://user-images.githubusercontent.com/64367553/106951540-f1f70e00-6740-11eb-8b56-324a757f6905.png)

## Example of usage
```py
class Set_Z_Index(Scene):
    def construct(self):
        line1 = Line(UP, DOWN, color=RED, z_index=1)
        line2 = Line(LEFT, RIGHT, color=GREEN, z_index=2)

        self.add(line1, line2)
        self.wait(0.1)
        self.play(line1.animate.set_z_index(3))
```

## What this code will do

![SetZIndex](https://user-images.githubusercontent.com/64367553/106955574-78fab500-6746-11eb-853e-467f8ccbda2b.gif)

## Questions to discuss
1. Is it a good idea to implement layers? Or just compare Mobject's by their z-indices?

>I don't think it would require anything more than sorting scene.mobjects by z-index before rendering each frame
and since its already possible to assign arbitrary attributes to mobjects it may end up being a one-line change
@eulertour 
2. What about performance if we will sort MObject's array by z-indices before each dt update? Maybe it's better to make observer that will look if one of scene objects changed its z-index and reorder only him? If so, how to do that?
3. What are your suggestions how to implement that feature?

## On the go
I really look forward for this feature. Unfortunately, I'm not strong enough to change something in Scene class. If someone help me with this implementation and prompt me what to do I can do PR.

**Comment by behackl** [comment]
I am a bit confused by this issue. We currently already order Mobjects based on their `z_index` -- however, you are right, animating such a change of layer does not work currently.

How did you generate the gif above? This sort of fading effect would be nice, but even interpolating the `z_index` values between two mobjects wouldn't yield something like that; one mobject would just jump above the other as soon as the `z_index` of the interpolation reaches a certain value. That behavior could be changed in theory, but would probably require playing around in `Camera`.

Could you please make clear what this issue actually is about?

**Comment by ponmike92** [comment]
@behackl ty for reply!

> We currently already order Mobjects based on their z_index. 

Cool! I thought that it is not working attribute. I had to set objects in right order or use add_foreground_mobjects() method. When it was implemented?

> Could you please make clear what this issue actually is about?

This issue is about having ability to change order of scene mobjects by they z-index through animation. 
Currently, If you want to change z-order of object you can do it before or after animation. There are no built in animations for that purpose.

@behackl What about idea adding z_ordering.py file to animations folder?
classes could be..
```py
class ChangeZOrder(Animation):
    def __init__(self, vmobject, lag_ratio=0.5, **kwargs):
        # vmobject instantly changes its z-order.
        # if lag_ratio = 0.5 it will be done in the middle of animation
        # if lag_ratio = 0 it will be done right in start of animation and so on..
        pass


class ChangingZOrder(Animation):
    def __init__(self, vmobject, **kwargs):
        # vmobject gradually changes its opacity relative to other objects on scene
        # by the end of animation vmobject finally changes its z-order
        pass
```

What I wan't to achieve is to have ability to combine these animations with mob.animate.set_color() and mob.animate.set_opacity() animations. I submitted common issue about z-ordering in #670 and @leotrs suggested me how to solve my problem. But it is not as flexible as could be.
